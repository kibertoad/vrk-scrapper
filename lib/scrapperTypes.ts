export type CandidateVotes = {
  name: string
  party: string
  districtVotes: number
  mailVotes: number
  totalVotes: number
  percentFromValid: number
  percentFromAll: number
}

export type DistrictResults = {
  attrHeaders: readonly string[]
  candidateValues: CandidateVotes[]
}
