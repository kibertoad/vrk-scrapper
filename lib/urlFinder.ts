// const urlRegex = /\'?srcUrl=\'(.*)\'.html\'/

export function findAllUrls(text: string): readonly string[] {
  const result = text.match(/\?srcUrl(.*)\.html/gi) ?? []
  return result.map((srcUrl) => {
    return srcUrl.replace('?srcUrl=', '')
  })
}
