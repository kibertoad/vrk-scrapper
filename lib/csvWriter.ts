import { createObjectCsvWriter } from 'csv-writer'
import { CandidateVotes } from './scrapperTypes'
import { stat, mkdir } from 'fs/promises'

function ensureDirectoryExists(dir: string) {
  return stat(dir).catch(() => mkdir(dir, { recursive: true }))
}

export async function writeSingleMandateResults(
  districtName: string,
  attrHeaders: readonly string[],
  candidateData: CandidateVotes[]
) {
  const dir = 'singleMandateByDistrict'
  await ensureDirectoryExists(dir)

  const headers = [
    {
      id: 'name',
      title: attrHeaders[0].toUpperCase(),
    },
    {
      id: 'party',
      title: attrHeaders[1].toUpperCase(),
    },
    {
      id: 'districtVotes',
      title: attrHeaders[2].toUpperCase(),
    },
    {
      id: 'mailVotes',
      title: attrHeaders[3].toUpperCase(),
    },
    {
      id: 'totalVotes',
      title: attrHeaders[4].toUpperCase(),
    },
    {
      id: 'percentFromValid',
      title: attrHeaders[5].toUpperCase(),
    },
    {
      id: 'percentFromAll',
      title: attrHeaders[6].toUpperCase(),
    },
  ]

  const csvWriter = createObjectCsvWriter({
    path: `${dir}/${districtName}.csv`,
    header: headers,
  })

  await csvWriter.writeRecords(candidateData)
}
