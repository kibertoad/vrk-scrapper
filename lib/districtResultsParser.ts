import parse5, { DefaultTreeParentNode, Document } from 'parse5'
import { CandidateVotes, DistrictVotes } from './types/scrapperTypes'

const linebreakRegex = /\r\n|\n|\r/gm
const IS_INDEPENDENT_TEXT = 'Išsikėlė pat'
const IS_INDEPENDENT_TEXT_OUTPUT = 'Išsikėlė pats(-i)'

type CandidateColumns = {
  nameColumn: any
  partyValue: string
  votesValue: string
}

// Structure of this data is not very consistent become amount of cells is different for independent candidates
function extractCandidateCells<T = any[]>(cells: any[]): CandidateColumns[] {
  const result: CandidateColumns[] = []
  let nextCandidate: Partial<CandidateColumns> = {}

  let foundName = false
  let foundParty = false
  let isIndependent = false
  for (let cell of cells) {
    if (!foundName) {
      nextCandidate.nameColumn = cell
      foundName = true
      continue
    }
    if (!foundParty && cell.childNodes) {
      nextCandidate.partyValue = cell.childNodes[0].value
      foundParty = true
      continue
    }

    // For independent candidates that fact is included in votes element. yay for consistency
    if (!foundParty && cell.value?.includes?.(IS_INDEPENDENT_TEXT)) {
      nextCandidate.partyValue = IS_INDEPENDENT_TEXT_OUTPUT
      foundParty = true
      isIndependent = true
    }

    if (foundParty) {
      nextCandidate.votesValue = isIndependent
        ? cell.value.replace(IS_INDEPENDENT_TEXT, '')
        : cell.value

      foundName = false
      foundParty = false
      isIndependent = false
      result.push(nextCandidate as CandidateColumns)
      nextCandidate = {}
    }
  }
  return result
}

function chunkAny(arr: any[], chunkSize: number): any[] {
  const R = []
  for (let i = 0, len = arr.length; i < len; i += chunkSize) R.push(arr.slice(i, i + chunkSize))
  return R
}

function transformCandidateValues(candidateRowsChunk: CandidateColumns): CandidateVotes {
  const name = candidateRowsChunk.nameColumn.childNodes[0].value
  const party = candidateRowsChunk.partyValue
  const voteValues: string[] = candidateRowsChunk.votesValue
    .trim()
    .replace(/\t/g, ' ')
    .replace(/    /g, ' ')
    .replace(/  /g, ' ')
    .replace(/  /g, ' ')
    .replace(/  /g, ' ')
    .replace(/  /g, ' ')
    .replace(linebreakRegex, '')
    .split(' ')

  const [districtVotes, mailVotes, totalVotes, percentFromValid, percentFromAll] = voteValues

  return {
    name,
    party,
    districtVotes,
    mailVotes,
    totalVotes,
    percentFromValid,
    percentFromAll,
  }
}

export function parseDistrictResults(text: string): DistrictVotes {
  const sanitizedText = text
    .trim()
    .replace(/\&nbsp;/g, ' ')
    .replace(linebreakRegex, '')
  const districtHtml: DefaultTreeParentNode = parse5.parse(sanitizedText) as DefaultTreeParentNode
  // @ts-ignore
  const candidateRows = districtHtml.childNodes[0].childNodes[1].childNodes
  const headerRow = candidateRows.shift()
  const attrHeaders = headerRow.value
    .trim()
    .replace(linebreakRegex, '')
    // we don't need this parent value
    .replace('Paduotų balsų skaičius', '')
    .replace(/\t/g, ' ')
    .replace(/\s{3,8}/g, '  ')
    .split('  ')

  const candidateRowsChunksWithSummary = extractCandidateCells(candidateRows)
  const candidateRowsChunks = candidateRowsChunksWithSummary.slice(
    0,
    candidateRowsChunksWithSummary.length - 3
  )
  const summaryRows = candidateRowsChunksWithSummary.slice(
    candidateRowsChunksWithSummary.length - 3,
    candidateRowsChunksWithSummary.length
  )
  const candidateValues = candidateRowsChunks.map(transformCandidateValues)

  return {
    attrHeaders,
    candidateValues,
  }
}
