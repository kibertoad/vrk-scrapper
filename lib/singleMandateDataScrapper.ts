import puppeteer, { Browser } from 'puppeteer'
import { findAllUrls } from './utils/urlFinder'
import { parseDistrictResults } from './districtResultsParser'
import { writeSingleMandateResults } from './utils/csvWriter'
import { DistrictResults } from './types/scrapperTypes'

const VRK_ROOT_URL = 'https://rezultatai.vrk.lt/'
const SINGLE_MANDATE_URL = '?srcUrl=/rinkimai/1104/1/1746/rezultatai/lt/rezultataiVienm.html'

const DISTRICT_TABLE_SELECTOR = 'table.partydata:nth-child(11)'

const DISTRICT_NAME_SELECTOR = '#pub-content > div:nth-child(2) > h3:nth-child(1)'
const DISTRICT_RESULT_TABLE_SELECTOR = 'table.partydata:nth-child(11)'
const DISTRICT_RESULT_TABLE_SELECTOR_ALTERNATIVE = 'table.partydata:nth-child(12)'

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

async function getDistrictUrls(browser: Browser): Promise<readonly string[]> {
  const page = await browser.newPage()
  await page.goto(`${VRK_ROOT_URL}${SINGLE_MANDATE_URL}`)
  await page.waitForSelector(DISTRICT_TABLE_SELECTOR)
  const districtTableHTML = await page.$eval(DISTRICT_TABLE_SELECTOR, (el) => {
    return el.innerHTML
  })
  const districtUrls = findAllUrls(districtTableHTML)
  return districtUrls
}

async function getDistrictResults(browser: Browser, districtUrl: string): Promise<DistrictResults> {
  const page = await browser.newPage()
  await page.goto(`${VRK_ROOT_URL}${districtUrl}`)
  await page.waitForSelector(DISTRICT_NAME_SELECTOR)

  const districtName = await page.$eval(DISTRICT_NAME_SELECTOR, (el) => {
    return el.innerHTML.replace('&nbsp;', ' ')
  })
  console.log(`Processing ${districtName}`)

  let resultsTableHTML
  try {
    resultsTableHTML = await page.$eval(DISTRICT_RESULT_TABLE_SELECTOR, (el) => {
      return el.innerHTML
    })
  } catch (err) {
    // Page structure is inconsistent across different districts, we need to handle that somehow
    resultsTableHTML = await page.$eval(DISTRICT_RESULT_TABLE_SELECTOR_ALTERNATIVE, (el) => {
      return el.innerHTML
    })
  }

  const districtResults = parseDistrictResults(resultsTableHTML)

  return {
    name: districtName,
    votes: districtResults,
  }
}

export async function scrapeSingleMandateData(): Promise<void> {
  const browser = await puppeteer.launch()

  try {
    const districtUrls = await getDistrictUrls(browser)

    for (const districtUrl of districtUrls) {
          const districtResults = await getDistrictResults(browser, districtUrl)
    // const districtResults = await getDistrictResults(browser, districtUrls[18])
    await writeSingleMandateResults(
      districtResults.name,
      districtResults.votes.attrHeaders,
      districtResults.votes.candidateValues
    )
    console.log(`Finished outputting district: ${districtResults.name}\n`)
    console.log('Sleeping...')
    await sleep(2000 + Math.random() * 1000)
    console.log('Woke up...')
    }
  } finally {
    await browser.close()
  }
}
