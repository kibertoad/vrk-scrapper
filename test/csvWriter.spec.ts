import fs from 'fs'
import path from 'path'
import { parseDistrictResults } from '../lib/districtResultsParser'
import { writeSingleMandateResults } from '../lib/csvWriter'
import { FileTestHelper } from 'cli-testlab'

const tableHTML = fs
  .readFileSync(path.resolve(__dirname, 'resources/single_mandate_district_results_table.html'))
  .toString()

describe('csvWriter', () => {
  it('writes single mandate results correctly', async () => {
    const fileHelper = new FileTestHelper(__dirname)

    const result = parseDistrictResults(tableHTML)
    await writeSingleMandateResults('district1', result.attrHeaders, result.candidateValues)
    const fileContent = fileHelper.getFileTextContent('../singleMandateByDistrict/district1.csv')
    expect(fileContent).toMatchSnapshot()

    await fileHelper.deleteDir('../singleMandateByDistrict')
  })
})
