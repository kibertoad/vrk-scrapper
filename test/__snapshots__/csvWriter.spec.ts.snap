// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvWriter writes single mandate results correctly 1`] = `
"KANDIDATAS,IŠKĖLĖ,APYLINKĖSE,PAŠTU,IŠ VISO,% NUO GALIOJANČIŲ BIULETENIŲ,% NUO DALYVAVUSIŲ RINKĖJŲ
Mykolas MAJAUSKAS,Tėvynės sąjunga – Lietuvos krikščionys demokratai,4465,1790,6255,\\"29,71\\",\\"28,92\\"
Viktorija ČMILYTĖ-NIELSEN,Lietuvos Respublikos liberalų sąjūdis,4582,1467,6049,\\"28,73\\",\\"27,97\\"
Vytenis ŠKARNULIS,Laisvės partija,1543,623,2166,\\"10,29\\",\\"10,02\\"
Darius KUOLYS,Politinė partija „Lietuvos sąrašas“,826,217,1043,\\"4,95\\",\\"4,82\\"
Dovilė ŠAKALIENĖ,Lietuvos socialdemokratų partija,767,254,1021,\\"4,85\\",\\"4,72\\"
Vitalija JANKAUSKAITĖ-MILČIUVIENĖ,Lietuvos valstiečių ir žaliųjų sąjunga,711,239,950,\\"4,51\\",\\"4,39\\"
Kristina DZERŽINSKA,Lietuvos lenkų rinkimų akcija - Krikščioniškų šeimų sąjunga,526,75,601,\\"2,85\\",\\"2,78\\"
Rasa IVANAUSKIENĖ,Darbo partija,461,126,587,\\"2,79\\",\\"2,71\\"
Zenonas VAIGAUSKAS,Lietuvos socialdemokratų darbo partija,415,118,533,\\"2,53\\",\\"2,46\\"
Robertas RAMANAUSKAS,Nacionalinis susivienijimas,409,90,499,\\"2,37\\",\\"2,31\\"
Diana STOMIENĖ,Partija „Laisvė ir teisingumas“,265,89,354,\\"1,68\\",\\"1,64\\"
Inga LAURUŠONĖ,Lietuvos žaliųjų partija,263,63,326,\\"1,55\\",\\"1,51\\"
Sakalas GORODECKIS,Centro partija - tautininkai,255,53,308,\\"1,46\\",\\"1,42\\"
Arvydas AKSTINAVIČIUS,Krikščionių sąjunga,165,58,223,\\"1,06\\",\\"1,03\\"
Andrius NIKITINAS,Partija LIETUVA – VISŲ,119,21,140,\\"0,66\\",\\"0,65\\"
"
`;
